File: HOT100/8.二叉树/124.二叉树中的最大路径和.py
#
# @lc app=leetcode.cn id=124 lang=python3
#
# [124] 二叉树中的最大路径和
#
from typing import Optional
class TreeNode:
    def __init__(self, _val, _left, _right):
        self.val = _val
        self.left = _left
        self.right = _right
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        maxVal = -float('inf')
        def dfs(n):
            if not n: return 0
            l = max(dfs(n.left), 0)
            r = max(dfs(n.right), 0)
            nonlocal maxVal
            maxVal = max(maxVal, l + r + n.val)
            return max(l, r) + n.val
        dfs(root)
        return maxVal
# @lc code=end

File: HOT100/8.二叉树/124.二叉树中的最大路径和.cpp
/*
 * @lc app=leetcode.cn id=124 lang=cpp
 *
 * [124] 二叉树中的最大路径和
 */
#include "tools.h"
using namespace std;
struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int res = INT_MIN;
        auto dfs = [&](auto&& self, TreeNode* n)
        {
            if (! n) return 0;
            int l = max(0, self(self, n->left));
            int r = max(0, self(self, n->right));
            res = max(res, n->val + l + r);
            return max(l, r) + n->val;
        };
        dfs(dfs, root);
        return res;
    }
};
// @lc code=end

